### PR TITLE
Fix DeepSpeedPlugin with IterableDataset

### PR DESCRIPTION
**Pull request by leezu**
Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/7345

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 1]
```diff
@@ -88,6 +88,7 @@ def __init__(
         allgather_bucket_size: int = 2e8,
         reduce_bucket_size: int = 2e8,
         zero_allow_untested_optimizer: bool = True,
+        logging_batch_size_per_gpu: Optional[int] = 1,
```
Comment (kaushikb11) [inline pytorch_lightning/plugins/training_type/deepspeed.py:91]:
```suggestion
        logging_batch_size_per_gpu: int = 1,
```

Comment (leezu) [reply pytorch_lightning/plugins/training_type/deepspeed.py:91]: If set to None, the logging_batch_size_per_gpu is inferred from the train DataLoader's BatchSampler

Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:91]: How about we follow the paradigm of using "auto" as infer?
```suggestion
        logging_batch_size_per_gpu: Union[int, str] = 1
```

**Comment by SeanNaren** [comment]
Thanks @leezu! Let me know your thoughts on the proposed change, and I can help make a test for this to get this merged.

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 1 above)
Comment (leezu) [reply pytorch_lightning/plugins/training_type/deepspeed.py:91]: Thanks @SeanNaren. Using `"auto"` sounds good to me and will result in easier to understand config files.

**Comment by SeanNaren** [comment]
I added a test, and modified the code to now use `auto`. Auto will try to use the batch sampler if it exists else will default to 1.

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 2]
```diff
@@ -409,14 +419,22 @@ def _format_batch_size_and_grad_accum_config(self):
...
             self.config["gradient_clipping"] = self.lightning_module.trainer.gradient_clip_val
 
+    def _auto_select_batch_size(self):
+        # train_micro_batch_size_per_gpu is used for throughput logging purposes
+        # by default we try to use the batch size of the loader
+        batch_size = 1
+        if hasattr(self.lightning_module, 'train_dataloader'):
+            train_dataloader = self.lightning_module.train_dataloader()
+            if hasattr(train_dataloader, 'batch_sampler'):
+                batch_size = train_dataloader.batch_sampler.batch_size
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/deepspeed.py:435]: If anyone can suggest anything cleaner please do :)

Comment (Borda) [reply pytorch_lightning/plugins/training_type/deepspeed.py:435]: probably not much as `train_dataloader()` is callable not just an attribute...

**Comment by Borda** [comment]
@leezu pls use and follow the bullet list from template

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/deepspeed.py:435]: This could break if the user provides several dataloaders to the CombinedLoader.

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by leezu** [comment]
Thank you @SeanNaren. LGTM

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 2 above)
Comment (Borda) [reply pytorch_lightning/plugins/training_type/deepspeed.py:435]: @leezu ^^

**Comment by SeanNaren** [comment]
Thanks @leezu :)
